### PR TITLE
Update Cobra from 1.2.1 to 1.3.0

### LINKED ISSUE #1166: Update Cobra to version 1.3.0
**What would you like to be added**:
Upgrade cobra from version 1.2.1 to 1.3.0

**Why is this needed**:
* Enables displaying command descriptions in bash completion output like it does for zsh and fish completions currently (bug fixed in cobra 1.3.0 to make this usable https://github.com/spf13/cobra/pull/1509)
* Improves flag parsing capabilities which will help with plugin command completions (https://github.com/kubernetes/kubernetes/pull/105867#issuecomment-995264504)
* Several other minor fixes and enhancements in Cobra 1.3.0 related to completions

### PR DESCRIPTION
**Pull request by brianpursley**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
This PR upgrades cobra from version 1.2.1 to [1.3.0](https://github.com/spf13/cobra/releases/tag/v1.3.0).

This will be helpful for kubectl for a few reasons:
* Enables displaying command descriptions in bash completion output like it does for zsh and fish completions currently (bug fixed in cobra 1.3.0 to make this usable https://github.com/spf13/cobra/pull/1509)
* Improves flag parsing capabilities which will help with plugin command completions (https://github.com/kubernetes/kubernetes/pull/105867#issuecomment-995264504)
* Several other minor fixes and enhancements in Cobra 1.3.0 related to completions

#### Which issue(s) this PR fixes:

Fixes https://github.com/kubernetes/kubectl/issues/1166

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

**Comment by marckhouzam** [comment]
Thanks @brianpursley for doing this!  I was trying to also do it but I had no idea how. 

What I tried to do myself is:
1. change the `spf13/cobra` and `spf13/viper` versions in all go.mod files
2. run `hack/update-vendor.sh` (a bit random but seemed appropriate 😄 )

This produced a much smaller diff than this PR and mine did not seem to have updated the vendor directory or the `go.mod` and `go.sum` files completely. But my `kubectl` built properly and did contain the new Cobra code.

 Is there a documentation on how this should be done?  Maybe I can help next time.

Either way, thanks for moving this forward.

**Comment by brianpursley** [comment]
> Is there a documentation on how this should be done? Maybe I can help next time.

@marckhouzam Eddie pointed me to [this documentation](https://github.com/kubernetes/community/blob/master/contributors/devel/sig-architecture/vendor.md), which I followed.  I was surprised at the number of dependencies which ended up needing to be updated, but I think it is due to cobra's dependency on a newer version of viper, which has a large number of dependencies.

**Comment by eddiezane** [comment]
Dep graph https://kuberneddies.s3.us-west-2.amazonaws.com/cobra-107628.svg.

`github.com/armon/go-metrics` appears to be the source of transitive additions. Brought in through viper.

Only physical license change to vendor is removal of `github.com/shurcooL/sanitized_anchor_name`

/assign @liggitt

It's your favorite time :smile:

File: go.mod [hunk 1]
```diff
@@ -466,17 +465,18 @@ replace (
 	gonum.org/v1/gonum => gonum.org/v1/gonum v0.6.2
 	gonum.org/v1/netlib => gonum.org/v1/netlib v0.0.0-20190331212654-76723241ea4e
 	gonum.org/v1/plot => gonum.org/v1/plot v0.0.0-20190515093506-e2840ee46a6b
-	google.golang.org/api => google.golang.org/api v0.46.0
```
Comment (eddiezane) [inline go.mod:469]: Big jump here.

File: go.mod [hunk 2]
```diff
@@ -158,44 +158,44 @@ replace (
...
 	github.com/PuerkitoBio/purell => github.com/PuerkitoBio/purell v1.1.1
 	github.com/PuerkitoBio/urlesc => github.com/PuerkitoBio/urlesc v0.0.0-20170810143723-de5bf2ad4578
 	github.com/ajstarks/svgo => github.com/ajstarks/svgo v0.0.0-20180226025133-644b8db467af
 	github.com/antihax/optional => github.com/antihax/optional v1.0.0
 	github.com/antlr/antlr4/runtime/Go/antlr => github.com/antlr/antlr4/runtime/Go/antlr v0.0.0-20210826220005-b48c857c3a0e
 	github.com/armon/circbuf => github.com/armon/circbuf v0.0.0-20150827004946-bbbad097214e
-	github.com/armon/go-metrics => github.com/armon/go-metrics v0.0.0-20180917152333-f0300d1749da
-	github.com/armon/go-radix => github.com/armon/go-radix v0.0.0-20180808171621-7fddfc383310
+	github.com/armon/go-metrics => github.com/armon/go-metrics v0.3.10
+	github.com/armon/go-radix => github.com/armon/go-radix v1.0.0
```
Comment (eddiezane) [inline go.mod:175]: Looks like this was just a pin https://github.com/armon/go-radix/releases/tag/v1.0.0

**Comment by liggitt** [comment]
wow, big dep change for a cli library bump 😮 

queued up for review

**Comment by liggitt** [comment]
> I think it is due to cobra's dependency on a newer version of viper, which has a large number of dependencies.

yeah... that might be something to keep on the back burner to investigate a way to isolate / drop ...

File: go.mod [hunk 3]
```diff
@@ -288,27 +288,28 @@ replace (
...
-	github.com/hashicorp/go-cleanhttp => github.com/hashicorp/go-cleanhttp v0.5.1
-	github.com/hashicorp/go-immutable-radix => github.com/hashicorp/go-immutable-radix v1.0.0
+	github.com/hashicorp/go-cleanhttp => github.com/hashicorp/go-cleanhttp v0.5.2
+	github.com/hashicorp/go-hclog => github.com/hashicorp/go-hclog v1.0.0
+	github.com/hashicorp/go-immutable-radix => github.com/hashicorp/go-immutable-radix v1.3.1
 	github.com/hashicorp/go-msgpack => github.com/hashicorp/go-msgpack v0.5.3
-	github.com/hashicorp/go-multierror => github.com/hashicorp/go-multierror v1.0.0
-	github.com/hashicorp/go-rootcerts => github.com/hashicorp/go-rootcerts v1.0.0
+	github.com/hashicorp/go-multierror => github.com/hashicorp/go-multierror v1.1.0
+	github.com/hashicorp/go-retryablehttp => github.com/hashicorp/go-retryablehttp v0.5.3
```
Comment (liggitt) [inline go.mod:299]: from github.com/circonus-labs/circonus-gometrics

File: go.mod [hunk 4]
```diff
@@ -332,22 +332,21 @@ replace (
 	github.com/liggitt/tabwriter => github.com/liggitt/tabwriter v0.0.0-20181228230101-89fcab3d43de
 	github.com/lithammer/dedent => github.com/lithammer/dedent v1.1.0
 	github.com/lpabon/godbc => github.com/lpabon/godbc v0.1.1
+	github.com/lyft/protoc-gen-star => github.com/lyft/protoc-gen-star v0.5.3
```
Comment (liggitt) [inline go.mod:335]: what is pulling this in? another proto generation tool increases likelihood of selecting incompatible transitive protobuf libraries

File: go.mod [hunk 5]
```diff
@@ -182,22 +182,22 @@ replace (
...
 	github.com/certifi/gocertifi => github.com/certifi/gocertifi v0.0.0-20200922220541-2c3bb06c6054
-	github.com/cespare/xxhash => github.com/cespare/xxhash v1.1.0
 	github.com/cespare/xxhash/v2 => github.com/cespare/xxhash/v2 v2.1.2
 	github.com/chai2010/gettext-go => github.com/chai2010/gettext-go v0.0.0-20160711120539-c6fed771bfd5
 	github.com/checkpoint-restore/go-criu/v5 => github.com/checkpoint-restore/go-criu/v5 v5.0.0
 	github.com/chzyer/logex => github.com/chzyer/logex v1.1.10
 	github.com/chzyer/readline => github.com/chzyer/readline v0.0.0-20180603132655-2972be24d48e
 	github.com/chzyer/test => github.com/chzyer/test v0.0.0-20180213035817-a1ea475d72b1
 	github.com/cilium/ebpf => github.com/cilium/ebpf v0.6.2
+	github.com/circonus-labs/circonus-gometrics => github.com/circonus-labs/circonus-gometrics v2.3.1+incompatible
```
Comment (liggitt) [inline go.mod:196]: what is pulling this in?

new transitive deps from this once we get to v3:
* https://github.com/circonus-labs/go-apiclient/blob/master/go.mod
* https://github.com/circonus-labs/circonus-gometrics/blob/master/go.mod
* https://github.com/circonus-labs/circonusllhist/blob/master/go.mod
* https://github.com/hashicorp/go-retryablehttp/blob/master/go.mod
* https://github.com/tv42/httpunix/blob/master/go.mod
* https://github.com/hashicorp/go-hclog/blob/master/go.mod

File: go.mod [hunk 6]
```diff
@@ -288,27 +288,28 @@ replace (
...
 	github.com/grpc-ecosystem/grpc-gateway => github.com/grpc-ecosystem/grpc-gateway v1.16.0
-	github.com/hashicorp/consul/api => github.com/hashicorp/consul/api v1.1.0
-	github.com/hashicorp/consul/sdk => github.com/hashicorp/consul/sdk v0.1.1
+	github.com/hashicorp/consul/api => github.com/hashicorp/consul/api v1.11.0
+	github.com/hashicorp/consul/sdk => github.com/hashicorp/consul/sdk v0.8.0
 	github.com/hashicorp/errwrap => github.com/hashicorp/errwrap v1.0.0
-	github.com/hashicorp/go-cleanhttp => github.com/hashicorp/go-cleanhttp v0.5.1
-	github.com/hashicorp/go-immutable-radix => github.com/hashicorp/go-immutable-radix v1.0.0
+	github.com/hashicorp/go-cleanhttp => github.com/hashicorp/go-cleanhttp v0.5.2
+	github.com/hashicorp/go-hclog => github.com/hashicorp/go-hclog v1.0.0
```
Comment (liggitt) [inline go.mod:295]: new

File: go.mod (hunk 3 above)
Comment (liggitt) [inline go.mod:299]: new

File: go.mod [hunk 7]
```diff
@@ -288,27 +288,28 @@ replace (
...
-	github.com/hashicorp/go-multierror => github.com/hashicorp/go-multierror v1.0.0
-	github.com/hashicorp/go-rootcerts => github.com/hashicorp/go-rootcerts v1.0.0
+	github.com/hashicorp/go-multierror => github.com/hashicorp/go-multierror v1.1.0
+	github.com/hashicorp/go-retryablehttp => github.com/hashicorp/go-retryablehttp v0.5.3
+	github.com/hashicorp/go-rootcerts => github.com/hashicorp/go-rootcerts v1.0.2
 	github.com/hashicorp/go-sockaddr => github.com/hashicorp/go-sockaddr v1.0.0
 	github.com/hashicorp/go-syslog => github.com/hashicorp/go-syslog v1.0.0
 	github.com/hashicorp/go-uuid => github.com/hashicorp/go-uuid v1.0.1
-	github.com/hashicorp/go.net => github.com/hashicorp/go.net v0.0.1
-	github.com/hashicorp/golang-lru => github.com/hashicorp/golang-lru v0.5.0
```
Comment (liggitt) [inline go.mod:303]: dropped

File: go.mod [hunk 8]
```diff
@@ -182,22 +182,22 @@ replace (
 	github.com/beorn7/perks => github.com/beorn7/perks v1.0.1
 	github.com/bgentry/speakeasy => github.com/bgentry/speakeasy v0.1.0
 	github.com/bits-and-blooms/bitset => github.com/bits-and-blooms/bitset v1.2.0
-	github.com/bketelsen/crypt => github.com/bketelsen/crypt v0.0.4
```
Comment (liggitt) [inline go.mod:185]: not dropped, switched to github.com/sagikazarmark/crypt fork

File: go.mod [hunk 9]
```diff
@@ -468,17 +467,18 @@ replace (
...
 	gonum.org/v1/netlib => gonum.org/v1/netlib v0.0.0-20190331212654-76723241ea4e
 	gonum.org/v1/plot => gonum.org/v1/plot v0.0.0-20190515093506-e2840ee46a6b
-	google.golang.org/api => google.golang.org/api v0.46.0
+	google.golang.org/api => google.golang.org/api v0.62.0
 	google.golang.org/appengine => google.golang.org/appengine v1.6.7
-	google.golang.org/genproto => google.golang.org/genproto v0.0.0-20210831024726-fe130286e0e2
-	google.golang.org/grpc => google.golang.org/grpc v1.40.0
+	google.golang.org/genproto => google.golang.org/genproto v0.0.0-20211208223120-3a66f561d7aa
+	google.golang.org/grpc => google.golang.org/grpc v1.42.0
+	google.golang.org/grpc/cmd/protoc-gen-go-grpc => google.golang.org/grpc/cmd/protoc-gen-go-grpc v1.1.0
```
Comment (liggitt) [inline go.mod:474]: same question about what is pulling this in?

**Comment by marckhouzam** [comment]
> > I think it is due to cobra's dependency on a newer version of viper, which has a large number of dependencies.
> 
> yeah... that might be something to keep on the back burner to investigate a way to isolate / drop ...

So kubernetes doesn't use Viper directly correct? And removing the dependency to Viper would have benefits for kubernetes?

If so, I will discuss this with the Cobra maintainers and see if there's anything that can be done.

File: go.mod (hunk 4 above)
Comment (eddiezane) [reply go.mod:335]: This appears to makes it's way in from [google.golang.org/grpc](https://github.com/grpc/grpc-go/blob/b4c5e24c7373b22c6c38d68379c39ecdc616a968/go.mod#L9) via [github.com/envoyproxy/go-control-plane](https://github.com/envoyproxy/go-control-plane/blob/main/go.mod#L8) via [github.com/envoyproxy/protoc-gen-validate](https://github.com/envoyproxy/protoc-gen-validate/blob/79071f0f8b04188b297a0517a6e55b2d3641ab5a/go.mod#L7).

**Comment by eddiezane** [comment]
> > > I think it is due to cobra's dependency on a newer version of viper, which has a large number of dependencies.
> > 
> > 
> > yeah... that might be something to keep on the back burner to investigate a way to isolate / drop ...
> 
> So kubernetes doesn't use Viper directly correct? And removing the dependency to Viper would have benefits for kubernetes?
> 
> If so, I will discuss this with the Cobra maintainers and see if there's anything that can be done.

If that could happen it would cut down on a ton of transitive deps for us.

ref: https://github.com/kubernetes/kubernetes/pull/102598

**Comment by liggitt** [comment]
Yeah, we dropped direct dependence on Viper in that PR, but we still have a transitive dependency on it, which adds a *lot* to our dependency graph. https://github.com/muesli/coral is an interesting proof-of-concept that tracks head of cobra and drops the viper dep. Since some of our deps use cobra though, I think the viper dependency would have to be made optional in cobra itself for it to benefit us:

```
go mod graph | grep " github.com/spf13/cobra" | grep -v k8s.io
github.com/coredns/corefile-migration@v1.0.14 github.com/spf13/cobra@v1.1.3
github.com/containerd/continuity@v0.1.0 github.com/spf13/cobra@v1.0.0
go.etcd.io/etcd/server/v3@v3.5.0 github.com/spf13/cobra@v1.1.3
go.etcd.io/etcd/pkg/v3@v3.5.0 github.com/spf13/cobra@v1.1.3
```

**Comment by jpmcb** [comment]
Hi all - I was pointed to this PR from some people in the cobra community. From what I gather, we need to remove the transient dependencies on Viper so that the dependencies resolve correctly?

**Comment by liggitt** [comment]
> The dependencies are resolving correctly, but are steadily increasing beyond acceptable levels because of the viper transitive dependency.

I'll be filing an issue in the cobra repo with a request to isolate the viper dependencies used only by the cobra command, with some suggestions/considerations based on our experience isolating dependencies for the cadvisor command (https://github.com/google/cadvisor/pull/2437)

**Comment by liggitt** [comment]
> I'll be filing an issue in the cobra repo with a request to isolate the viper dependencies used only by the cobra command, with some suggestions/considerations based on our experience isolating dependencies for the cadvisor command ([google/cadvisor#2437](https://github.com/google/cadvisor/pull/2437))

opened https://github.com/spf13/cobra/issues/1597

**Comment by marckhouzam** [comment]
Do you guys still plan to go ahead with upgrading to Cobra 1.3 or do you want to wait for the next release of Cobra which will remove the dependency on Viper?  
Some Cobra 1.3 features could be used right away (completion description for Bash), and others would unblock other PRs (completion for plugins #105867).

**Comment by dims** [comment]
@marckhouzam i'd rather wait until viper is out.

**Comment by marckhouzam** [comment]
> FYI, the viper dependency has been dropped from cobra HEAD in [spf13/cobra#1604](https://github.com/spf13/cobra/pull/1604), and cobra v1.4.0 will be tagged shortly. Once it is, this can be updated to cobra v1.4.0 🎉

Thanks @liggitt for working with Cobra to make this happen.  It will benefit a whole bunch of projects 🎉

**Comment by dims** [comment]
thanks a ton @liggitt @jpmcb !!

**Comment by brianpursley** [comment]
Closing this PR.

It sounds like @liggitt may open a PR to upgrade to Cobra 1.4.0.

**Pull request closed without merging**